src/metabase/metabot/util.clj

```diff
@@ -128,16 +148,51 @@
 (defn- add-create-table-ddl [model]
   (assoc model :create_table_ddl (create-table-ddl model)))
 
+(defn- add-sql-names
+  "Add a distinct SCREAMING_SNAKE_CASE sql name to each field in the result_metadata."
+  [{:keys [result_metadata] :as model}]
+  (let [m (->> result_metadata
+               (map
+                 (fn [{:keys [display_name] :as rsmd}]
+                   ;; Add in a baseline name
+                   (assoc rsmd :sql_name (normalize-name display_name))))
+               ;; Sort name length (desc) then sql_name.
+               ;; We want longer names first so pre-existing conflicts will resolve first
+               ;; e.g. ABC, ABC, ABC_0 should result in ABC_0, ABC, ABC_1, not ABC, ABC_0, ABC_0_0
+               ;; The remaining sort keys beyond :sql_name aren't that important.
+               ;; They are just there for stability.
+               (sort-by (juxt (comp - count :sql_name)
+                              :sql_name
+                              :display_name
+                              :name
+                              :id)))]
+    (loop [[{:keys [sql_name] :as f} & r] m
```

Comment:
I think this is way more complicated than it needs to be. We have a function `metabase.mbql.util/unique-name-generator` that can help out (and i'm surprised the alias middleware didn't use it) that should solve all this for you. We can merge this as is, but in the future we should have a function that takes a seq of names and returns a zipmap of names to disambiguated names. The fact that this function is intertwined with result_metadata and sql_name i think makes this way more complicated.

Something like

```clojure
(defn disambiguate [names]
  (let [uniquifier    (metabase.mbql.util/unique-name-generator)
        [_ new-names] (reduce (fn [[taken acc] n]
                                (let [candidate (uniquifier n)]
                                  (if (taken candidate)
                                    (recur [(conj taken candidate) acc] n)
                                    [(conj taken candidate) (conj acc candidate)])))
                              [#{} []]
                              names)]
    (map vector names new-names)))

(disambiguate ["x" "y" "x" "x_1" "x_2"])
(["x" "x"] ["y" "y"] ["x" "x_2"] ["x_1" "x_1"] ["x_2" "x_2_2"])
```

(can also let it "see" ahead to the names not yet seen, etc. But just pulling this logic out of the "domain" (result_metadata, formatting aliases, etc) can make the task much simpler, and then assemble it where needed.

Reply:
Done.